api/controllers/AuthController.js

```diff
@@ -25,49 +25,67 @@ module.exports = {
-          /* Redirect to the mod authentication page, or to the desired url if this was mod authentication.*/
-          if (login_info.type !== 'mod') {
-            return res.redirect('/auth/reddit?loginType=mod' + (login_info.redirect ? '&redirect=' + encodeURIComponent(login_info.redirect) : ''));
-          }
-          finishLogin();
-        });
-      }
-      else if (user.isMod) { // User is not a mod, but had isMod set for some reason (e.g. maybe the user used to be a mod). Set isMod to false.
-        User.update(user.name, {isMod: false}).exec(finishLogin);
-      } else { // Regular user
-        finishLogin();
+          return res.view(403, {error: 'Sorry, something went wrong. Try logging in again.'});
+        }
+        var login_info;
+        try {
+          login_info = JSON.parse(req.query.state);
+        } catch (err) {
+          console.log('Error with parsing /u/' + user.name + '\'s session state');
+          return res.serverError(err);
+        }
+        if (login_info.validation !== req.session.validation) {
+          console.log("Failed login for /u/" + user.name + ": invalid session state");
+          return res.view(403, {error: 'You have an invalid session state. (Try logging in again.)'});
+        }
+        let finishLogin = function () {
+          req.logIn(user, function (err) {
+            if (err) {
+              sails.error('Failed login: ' + err);
+              return res.forbidden(err);
+            }
+            var url = decodeURIComponent(login_info.redirect);
+            // Don't redirect to other callback urls (this may cause infinite loops) or to absolute url paths (which might lead to other sites).
+            if (url.indexOf('/auth/reddit/callback') === 0 || /^(?:[a-z]+:)?\/\//i.test(url)) {
+              url = '/';
+            }
+            req.session.validation = '';
+            return res.redirect(url);
+          });
+        };
+        let modStatus = await Reddit.checkModeratorStatus(sails.config.reddit.adminRefreshToken, user.name, 'pokemontrades');
+        if (modStatus) { //User is a mod, set isMod to true
+          User.update(user.name, {isMod: true}).exec(function () {
+            /* Redirect to the mod authentication page, or to the desired url if this was mod authentication.*/
+            if (login_info.type !== 'mod') {
+              return res.redirect('/auth/reddit?loginType=mod' + (login_info.redirect ? '&redirect=' + encodeURIComponent(login_info.redirect) : ''));
+            }
+            return finishLogin();
+          });
+        }
+        else if (user.isMod) { // User is not a mod, but had isMod set for some reason (e.g. maybe the user used to be a mod). Set isMod to false.
+          User.update(user.name, {isMod: false}).exec(finishLogin);
+        } else { // Regular user
+          return finishLogin();
+        }
+      } catch (err) {
+        return res.serverError(err);
```

Comment:
I put this function in a try/catch since it's async, so that accounts for most of this diff.